app/services/middleware_topology_service.rb

```diff
@@ -0,0 +1,69 @@
+class MiddlewareTopologyService
+  def initialize(provider_id)
+    @provider_id = provider_id
+    @providers = retrieve_providers
+  end
+
+  def build_topology
+    topology = {}
+    topo_items = {}
+    links = []
+
+    @providers.each do |provider|
+      topo_items[provider.id.to_s] = build_entity_data(provider, provider.type.split('::')[2])
+      provider.middleware_servers.each do |n|
+        topo_items[n.ems_ref] = build_entity_data(n, "MiddlewareServer")
+        links << build_link(provider.id.to_s, n.ems_ref)
+        n.middleware_deployments.each do |cg|
+          topo_items[cg.ems_ref] = build_entity_data(cg, "MiddlewareDeployment")
+          links << build_link(n.ems_ref, cg.ems_ref)
+        end
+      end
+    end
+
+    topology[:items] = topo_items
+    topology[:relations] = links
+    topology[:kinds] = build_kinds
+    topology
+  end
+
+  def build_entity_data(entity, kind)
+    status = entity_status(entity, kind)
+    id = case kind
+         when 'MiddlewareDeployment', 'MiddlewareServer' then entity.nativeid
+         when 'Hawkular' then entity.id.to_s
+         else entity.ems_ref
```

Comment:
I think it's weird to use these different identifiers all over here.  We already have a strategy to create simple node ids for the UI (cc @dclarizio, @chessbyte), so it should use the same one.  This is probably more of a discussion with @abonas.